contracts/src/RedemptionHelper.sol

```diff
@@ -0,0 +1,165 @@
+
+                LatestTroveData memory trove = ITroveManager(branch[i].troveManager).getLatestTroveData(troveId);
+                if (trove.entireColl * branch[i].price / trove.entireDebt >= _100pct) {
+                    branch[i].redeemedBold +=
+                        Math.min(branch[i].attemptedBold - branch[i].redeemedBold, trove.entireDebt);
+                }
+
+                troveId = nextTroveId;
+                nextTroveId = ISortedTroves(branch[i].sortedTroves).getPrev(nextTroveId);
+            }
+        }
+    }
+
+    // Meant to be called off-chain
+    // Not a view because price fetching has side-effects
+    function truncateRedemption(uint256 _bold, uint256 _maxIterationsPerCollateral)
+        external
+        returns (uint256 truncatedBold, uint256 feePct, Redeemed[] memory redeemed)
+    {
+        (SimulationContext[] memory branch, uint256 totalProportions) =
+            simulateRedemption(_bold, _maxIterationsPerCollateral);
+
+        if (totalProportions == 0) return (0, 0, redeemed);
+
+        truncatedBold = _bold;
+        for (uint256 i = 0; i < numBranches; ++i) {
+            if (branch[i].redeemable && branch[i].proportion > 0) {
+                uint256 extrapolatedBold = branch[i].redeemedBold * totalProportions / branch[i].proportion;
+                if (extrapolatedBold < truncatedBold) truncatedBold = extrapolatedBold;
+            }
+        }
+
+        feePct = collateralRegistry.getRedemptionRateForRedeemedAmount(truncatedBold);
+        redeemed = new Redeemed[](numBranches);
+
+        for (uint256 i = 0; i < numBranches; ++i) {
+            (uint256 redemptionPrice,) = addresses[i].priceFeed().fetchRedemptionPrice();
+            redeemed[i].bold = truncatedBold * branch[i].proportion / totalProportions;
+            redeemed[i].coll = redeemed[i].bold * (DECIMAL_PRECISION - feePct) / redemptionPrice;
+        }
+    }
+
+    function redeemCollateral(
+        uint256 _bold,
+        uint256 _maxIterationsPerCollateral,
+        uint256 _maxFeePct,
+        uint256[] memory _minCollRedeemed
+    ) external {
+        require(_minCollRedeemed.length == numBranches, "Wrong _minCollRedeemed length");
+
+        RedemptionContext[] memory branch = new RedemptionContext[](numBranches);
+
+        for (uint256 i = 0; i < numBranches; ++i) {
+            branch[i].collToken = collateralRegistry.getToken(i);
+            branch[i].collBalanceBefore = branch[i].collToken.balanceOf(address(this));
+        }
+
+        uint256 boldBalanceBefore = boldToken.balanceOf(address(this));
+
+        boldToken.transferFrom(msg.sender, address(this), _bold);
+        collateralRegistry.redeemCollateral(_bold, _maxIterationsPerCollateral, _maxFeePct);
+
+        for (uint256 i = 0; i < numBranches; ++i) {
+            branch[i].collRedeemed = branch[i].collToken.balanceOf(address(this)) - branch[i].collBalanceBefore;
```

Comment:
Why do we put `collRedeemed` in the branch struct array? As we are only using it inside this loop, can’t it just be a temporary variable here?